std/range/package.d

```diff
@@ -1447,97 +1448,121 @@ if (isInputRange!(Unqual!R1) && isInputRange!(Unqual!R2) &&
 }
 
 
-private struct ChooseResult(R1, R2)
+private struct ChooseResult(Ranges...)
 {
-    import std.traits : hasElaborateCopyConstructor, hasElaborateDestructor;
+    import std.meta : aliasSeqOf, ApplyLeft;
+    import std.traits : hasElaborateCopyConstructor, hasElaborateDestructor,
+        lvalueOf;
 
     private union
     {
-        R1 r1;
-        R2 r2;
+        Ranges rs;
     }
-    private bool r1Chosen;
+    private size_t chosenI;
 
-    private static auto ref actOnChosen(alias foo, ExtraArgs ...)(ref ChooseResult r,
-            auto ref ExtraArgs extraArgs)
+    private static auto ref actOnChosen(alias foo, ExtraArgs ...)
+        (ref ChooseResult r, auto ref ExtraArgs extraArgs)
     {
-        if (r.r1Chosen)
-        {
-            ref get1(return ref ChooseResult r) @trusted { return r.r1; }
-            return foo(get1(r), extraArgs);
-        }
-        else
+        ref getI(size_t i)(return ref ChooseResult r) @trusted { return r.rs[i]; }
+
+        switch(r.chosenI)
         {
-            ref get2(return ref ChooseResult r) @trusted { return r.r2; }
-            return foo(get2(r), extraArgs);
+            static foreach(candI; 0 .. rs.length)
+            {
+                case candI: return foo(getI!candI(r), extraArgs);
+            }
+
+            default: assert(false);
         }
     }
 
-    this(bool r1Chosen, return scope R1 r1, return scope R2 r2) @trusted
+    // @trusted because of assignment of r which overlap each other
+    this(size_t chosen, return scope Ranges rs) @trusted
     {
-        // @trusted because of assignment of r1 and r2 which overlap each other
         import core.lifetime : emplace;
 
-        // This should be the only place r1Chosen is ever assigned
+        // This should be the only place chosenI is ever assigned
         // independently
-        this.r1Chosen = r1Chosen;
-        if (r1Chosen)
+        this.chosenI = chosen;
+
+        // Otherwise the compiler will complain about skipping these fields
+        static foreach(i; 0 .. rs.length)
         {
-            this.r2 = R2.init;
-            emplace(&this.r1, r1);
+            this.rs[i] = Ranges[i].init;
         }
-        else
+
+        // The relevant field needs to be initialized last so it will overwrite
+        // the other initializations and not the other way around.
+        sw: switch(chosenI)
         {
-            this.r1 = R1.init;
-            emplace(&this.r2, r2);
+            static foreach(i; 0 .. rs.length)
+            {
+                case i:
+                this.rs[i] = rs[i];
```

Comment:
Given that you're initializing all union fields before this (why all, by the way?), this line would be assignment and not initialization, which changes semantics thus making this a breaking change. Consider keeping `emplace`?

Reply:
> you're initializing all union fields before this (why all, by the way?)
Because if i do `if (chosenI != i)` i get complains from the compiler that an initialization may be skipped. The only way I could avoid that is doing nested `static foreach`s, because the chosen range needs to be initialized last. I want to avoid O(n^^2) compile time complexity.

> this line would be assignment and not initialization, which changes semantics thus making this a breaking change

Good catch! My thought patterns were: Since I have already initialized that member, I can't `emplace` on it or I'm skipping it's potential destructor. So I initialize it normally. I forgot that:
1. assignment is different from initialization.
2. it's an union so it won't destroy the old value

Yes, I have to put the `emplace` back.